libcasr/src/stacktrace.rs

```diff
@@ -190,67 +190,68 @@ pub fn dedup_stacktraces(stacktraces: &[Stacktrace]) -> Vec<bool> {
 /// An vector of the same length as `stacktraces`.
 /// Vec\[i\] is the flat cluster number to which original stacktrace i belongs.
 pub fn cluster_stacktraces(stacktraces: &[Stacktrace]) -> Result<Vec<u32>> {
+    // Writing distance matrix
+    // Only the values in the upper triangle are explicitly represented,
+    // not including the diagonal
     let len = stacktraces.len();
-    // Writing compressed distance matrix into Vector<String>
-    let lines: Vec<String> = (0..len)
-        .map(|i| {
-            let mut tmp_str = String::new();
-            for j in i + 1..len {
-                tmp_str += format!(
-                    "{0:.3} ",
-                    1.0 - similarity(&stacktraces[i], &stacktraces[j])
-                )
-                .as_str();
-            }
-            tmp_str
-        })
-        .collect();
-
-    let python_cluster_script =
-        "import numpy as np;\
-        from scipy.cluster.hierarchy import fcluster, linkage;\
-        a = np.fromstring(input(), dtype=float, sep=' ');\
-        print(*fcluster(linkage([a] if type(a.tolist()) is float else a, method=\"complete\"), 0.3, criterion=\"distance\"))";
-
-    let Ok(mut python) = Command::new("python3")
-        .args(["-c", python_cluster_script])
-        .stdin(Stdio::piped())
-        .stdout(Stdio::piped())
-        .stderr(Stdio::piped())
-        .spawn() else {
-        return Err(Error::Casr("Failed to launch python3".to_string()));
-    };
-    {
-        let python_stdin = python.stdin.as_mut().unwrap();
-        if python_stdin.write_all(lines.join("").as_bytes()).is_err() {
-            return Err(Error::Casr(
-                "Error while writing to stdin of python script".to_string(),
-            ));
+    let size = len * (len - 1) / 2;
```

Comment:
I found this code snippet from [here](https://docs.rs/kodama/latest/kodama/index.html#example)
```rust
// Build our condensed matrix by computing the dissimilarity between all
// possible coordinate pairs.
let mut condensed = vec![];
for row in 0..coordinates.len() - 1 {
    for col in row + 1..coordinates.len() {
        condensed.push(haversine(coordinates[row], coordinates[col]));
    }
}
```
I suppose we can do the same way using push?